google/cloud/bigtable/internal/completion_queue_impl.h

```diff
@@ -62,6 +65,88 @@ class AsyncGrpcOperation : public AsyncOperation {
   virtual bool Notify(CompletionQueue& cq, bool ok) = 0;
 };
 
+/**
+ * Wrap a unary RPC callback into a `AsyncOperation`.
+ *
+ * This class is used by the implementation of `CompletionQueue` to associate
+ * a future with an asynchronous unary RPC call. gRPC requires applications to
+ * provide a `grpc::ClientContext` object, an object of the response type, and a
+ * `grpc::Status` object to make an asynchronous RPC. The lifetime of these
+ * objects must be at least as long as the duration of the asynchronous call.
+ * Furthermore, the application must provide a unique `void*` that is associated
+ * with the RPC.
+ *
+ * This class is used in the implementation of `CompletionQueue` to hold the
+ * objects mentioned above. Furthermore, when the operation is completed, it
+ * transfers the result to satisfy the future associated with the RPC.
+ *
+ * @tparam Request the type of the RPC request.
+ * @tparam Response the type of the RPC response.
+ */
+template <typename Request, typename Response>
+class AsyncUnaryRpcFuture : public AsyncGrpcOperation {
+ public:
+  explicit AsyncUnaryRpcFuture() : sync_(false) {}
+
+  future<StatusOr<Response>> GetFuture() { return promise_.get_future(); }
+
+  /// Prepare the operation to receive the response and start the RPC.
+  template <typename AsyncFunctionType>
+  void Start(AsyncFunctionType async_call,
+             std::unique_ptr<grpc::ClientContext> context,
+             Request const& request, grpc::CompletionQueue* cq, void* tag) {
+    context_ = std::move(context);
+    // Make sure context_ is visible in other threads.
+    sync_.store(0, std::memory_order_release);
+    auto rpc = async_call(context_.get(), request, cq);
+    rpc->Finish(&response_, &status_, tag);
+  }
+
+  void Cancel() override {
+    // Make sure context_ is visible in this thread.
+    static_cast<void>(sync_.load(std::memory_order_acquire));
+    context_->TryCancel();
+  }
+
+ private:
+  bool Notify(CompletionQueue&, bool ok) override {
+    // Make sure changes to the member variables are visible.
+    static_cast<void>(sync_.load(std::memory_order_acquire));
+    if (!ok) {
+      // This would mean a bug in gRPC. The documentation states that Finish()
+      // always returns `true` for unary RPCs.
+      promise_.set_value(::google::cloud::Status(
+          google::cloud::StatusCode::kUnknown, "Finish() returned false"));
+      return true;
+    }
+    if (!status_.ok()) {
+      // Convert the error to a `google::cloud::Status` and satisfy the future.
+      promise_.set_value(MakeStatusFromRpcError(status_));
+      return true;
+    }
+    // Success, use `response_` to satisfy the future.
+    promise_.set_value(std::move(response_));
+    return true;
+  }
+
+  // While only one thread changes (or reads) the state of this class at a time,
```

Comment:
As I recall @dopiera ran into actual problems (see #1420, #1419) before we decided to add these synchronization primitives.  I made the comment even more explicit, if that does not work we probably need to GVC or something.

Reply:
For #1419 at least, it sounds like ...

`response_reader_ = (client.*call)(context_.get(), request, cq, tag);`

might not finish the assignment until after the client call completes and `Notify()` is invoked, and that sounds like a flaw in that call interface.  If so, I'm happy to admit a mutex would avoid that because it would block the `Notify()` thread, but the atomic doesn't have that behavior.